### PR TITLE
Accept 0-dim channel inputs in convolution layer

### LINKED ISSUE #56998: Empty channel support in convolution
## 🚀 Feature
Convolution ops now support batch_size=0 (#12013). We'd like to let convolution ops support inputs/weights with channel=0.

## Motivation

We've encountered such a situation in detection research, where the #channels is dynamically generated per iteration (and could be 0). We have to workaround it with matmul instead:
```python
import torch
from torch.nn import functional as F

C = 8
D = 16  # Cout
K = 10


def f1(features, weights, bias):
    features = features.reshape(N, C, K)
    weights = weights.reshape(N, D, C)
    bias = bias.reshape(N, D, 1)
    out = torch.einsum("nck,ndc->ndk", features, weights) + bias
    return out.reshape(1, N * D, K)


def f2(features, weights, bias):
    return F.conv1d(features, weights, bias, stride=1, padding=0, groups=N)


for N in [3, 0]:
    features = torch.rand(1, N * C, K)
    weights = torch.rand(N * D, C, 1)
    bias = torch.rand(N * D)
    out1 = f1(features, weights, bias)
    out2 = f2(features, weights, bias)
    print(torch.abs(out1.reshape(out2.shape)-out2).max())
```

In the above code, `f2` and `f1` are equivalent except that `f2` segfaults when `N=0`. The goal is to make them truly equivalent.

### PR DESCRIPTION
**Pull request by v0dro**
Fixes https://github.com/pytorch/pytorch/issues/56998 .

### CODE REVIEW

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update & sorry for the reviewing delay! I added some comments below

**Review by jbschlosser — COMMENTED** [review]
Just made another pass! Flagged a couple things, summarized here:
1. `input` with 0 channels should error out when `weight` has non-zero channels (channels still have to match)
2. We want grad to exist for `weight` / `bias` even if empty tensors

Suggested some changes below that I believe will get us this :)

**Review by jbschlosser — COMMENTED** [review]
Thanks for the updates! It's looking pretty solid - good to go once we nail down the check message wording :)

**Review by jbschlosser — APPROVED** [review]
LGTM thanks! :)
